File: frontend/lib/frontier-carbon-compare.ts
/**
 * Rebuild frontier model comparison from workflow carbon_data.
 *
 * Each frontier bar = entire workflow (map+compile token mass + shared stages)
 * served by that single model. Our system = per-chunk routed estimate.
 */

import { A_J_PER_TOKEN_MEDIUM } from "./carbon-constants"

export const FRONTIER_MODEL_J_PER_TOKEN: ReadonlyArray<readonly [string, number]> = [
  ["GPT-o3", 7.5],
  ["GPT-4", 6.5],
  ["Llama 4 Behemoth", 6.5],
  ["Claude 4 Opus", 6.0],
  ["GPT-4o", 4.0],
  ["Gemini 2.5 Pro", 3.8],
  ["Llama 4 Maverick", 2.8],
  ["Gemma 3", 1.6],
] as const

/** @deprecated factor alias vs medium J — prefer FRONTIER_MODEL_J_PER_TOKEN */
export const FRONTIER_RELATIVE_INTENSITY: ReadonlyArray<readonly [string, number]> =
  FRONTIER_MODEL_J_PER_TOKEN.map(
    ([name, j]) => [name, Math.round((j / A_J_PER_TOKEN_MEDIUM) * 10000) / 10000] as const,
  )

export const OUR_SYSTEM_NAME = "Green Agentic Document Processing System"
export const OUR_SYSTEM_TAGLINE = "Smart Carbon-Aware Routing"

const LEGACY_CHUNK_GRAMS = 5.25
const LEGACY_BASELINE_GUARD = 150
const JOULES_PER_KWH = 3_600_000
const HEAVY_J = 6.5

function num(value: unknown, fallback = 0): number {
  const n = Number(value)
  return Number.isFinite(n) ? n : fallback
}

function round1(value: number): number {
  return Math.round(value * 10) / 10
}

function round4(value: number): number {
  return Math.round(value * 10000) / 10000
}

export type ChunkBreakdownRow = {
  chunk_index?: number
  tier?: string
  model?: string
  input_tokens?: number
  map_tokens?: number
  energy_kwh?: number
  co2e_g?: number
  j_per_token?: number
}

export type CarbonLike = {
  baseline_cost_gco2e?: number
  actual_cost_gco2e?: number
  carbon_saved_grams?: number
  efficiency_percent?: number
  emissions_direction?: string
  total_chunks?: number
  baseline_energy_kwh?: number
  actual_energy_kwh?: number
  local_grid_gco2_kwh?: number
  input_tokens?: number
  methodology?: string | null
  chunk_breakdown?: ChunkBreakdownRow[] | null
  breakdown?: {
    baseline_co2e_g?: number
    actual_co2e_g?: number
    baseline_energy_kwh?: number
    optimized_energy_kwh?: number
    grid_carbon_intensity_gco2_kwh?: number
    input_tokens?: number
    compile_tokens?: number
    map_tokens_total?: number
    map_tokens_by_tier?: Record<string, number>
    baseline_stages_gco2e?: Record<string, number>
    chunk_breakdown?: ChunkBreakdownRow[]
    emissions_direction?: string
    carbon_saved_g?: number
    reduction_percent?: number
    [key: string]: unknown
  } | null
}

export type ComparisonModelRow = {
  model: string
  relative_factor: number
  estimated_gco2e: number
  saved_gco2e: number
  reduction_percent: number
}

export type FrontierComparisonPayload = {
  comparison_models: ComparisonModelRow[]
  our_system: { name: string; tagline: string; carbon: number }
  summary_cards: {
    actual_emissions_gco2e: number
    carbon_saved_gco2e: number
    reduction_percent: number
    heavy_model_baseline_gco2e: number
    estimated_optimized_pipeline_emissions_g?: number
    estimated_baseline_pipeline_emissions_g?: number
    emissions_direction?: string
  }
  badges: string[]
  chart_bars: Array<{ model: string; estimated_gco2e: number; is_ours: boolean }>
  methodology: string
  chunk_breakdown?: ChunkBreakdownRow[]
}

function resolveBaseline(carbon: CarbonLike): number {
  const bd = carbon.breakdown || {}
  let baseline = num(carbon.baseline_cost_gco2e)
  if (baseline <= 0) baseline = num(bd.baseline_co2e_g)

  const energy =
    num(carbon.baseline_energy_kwh) || num(bd.baseline_energy_kwh)
  const intensity =
    num(carbon.local_grid_gco2_kwh) ||
    num(bd.grid_carbon_intensity_gco2_kwh)

  if (baseline > LEGACY_BASELINE_GUARD && energy > 0 && intensity > 0) {
    const rebuilt = energy * intensity
    if (rebuilt > 0 && rebuilt < baseline) baseline = rebuilt
  }

  if (baseline <= 0 && energy > 0 && intensity > 0) {
    baseline = energy * intensity
  }

  const chunks = num(carbon.total_chunks)
  if (
    baseline > LEGACY_BASELINE_GUARD &&
    chunks > 0 &&
    Math.abs(baseline - chunks * LEGACY_CHUNK_GRAMS) < 0.5
  ) {
    if (energy > 0 && intensity > 0) baseline = energy * intensity
  }

  return baseline
}

function workflowInferenceTokens(carbon: CarbonLike): number {
  const bd = carbon.breakdown || {}
  let mapTotal = num(bd.map_tokens_total)
  if (mapTotal <= 0 && bd.map_tokens_by_tier) {
    mapTotal = Object.values(bd.map_tokens_by_tier).reduce((s, v) => s + num(v), 0)
  }
  const compileTok = num(bd.compile_tokens)
  if (mapTotal + compileTok > 0) return mapTotal + compileTok
  const inp = num(bd.input_tokens || carbon.input_tokens)
  if (inp > 0) return Math.max(Math.floor(inp * 1.25), inp) + Math.floor(inp / 3)
  return 0
}

function estimateFrontierG(
  carbon: CarbonLike,
  baseline: number,
  modelJ: number,
): number {
  const bd = carbon.breakdown || {}
  const intensity =
    num(carbon.local_grid_gco2_kwh) ||
    num(bd.grid_carbon_intensity_gco2_kwh) ||
    700.0
  const tokens = workflowInferenceTokens(carbon)

  // Absolute estimate when baseline is missing (Summary Ready before full carbon).
  if (baseline <= 0) {
    if (tokens > 0 && intensity > 0) {
      return Math.max(0, ((tokens * modelJ) / JOULES_PER_KWH) * intensity)
    }
    return 0
  }

  const stages = (bd.baseline_stages_gco2e || {}) as Record<string, number>
  let inferenceG = num(stages.inference_gco2e)
  let otherG = 0
  for (const [k, v] of Object.entries(stages)) {
    if (k === "inference_gco2e" || k === "total_gco2e" || k === "infrastructure_gco2e") continue
    otherG += num(v)
  }
  let infraG = num(stages.infrastructure_gco2e)
  if (inferenceG <= 0) {
    inferenceG = baseline * 0.86
    otherG = baseline * 0.05
    infraG = Math.max(0, baseline - inferenceG - otherG)
  }

  let newInferenceG: number
  if (tokens > 0) {
    // IT-only (no PUE) — infrastructure share is rescaled below (matches backend).
    const itJ = tokens * modelJ
    newInferenceG = (itJ / JOULES_PER_KWH) * intensity
  } else {
    newInferenceG = inferenceG * (modelJ / HEAVY_J)
  }

  const itOld = otherG + inferenceG
  const itNew = otherG + newInferenceG
  const infraNew = itOld > 0 ? infraG * (itNew / itOld) : infraG
  return Math.max(0, itNew + infraNew)
}

export function looksLikeLegacyComparison(
  models: Array<{ estimated_gco2e?: number }> | null | undefined,
  carbon?: CarbonLike | null,
): boolean {
  if (!models?.length) return true
  const maxEst = Math.max(...models.map((m) => num(m.estimated_gco2e)))
  if (maxEst >= LEGACY_BASELINE_GUARD) return true
  const chunks = num(carbon?.total_chunks)
  if (chunks > 0 && Math.abs(maxEst - chunks * LEGACY_CHUNK_GRAMS) < 1) {
    return true
  }
  return false
}

export function buildFrontierComparisonFromCarbon(
  carbon: CarbonLike | null | undefined,
): FrontierComparisonPayload | null {
  if (!carbon) return null

  const baseline = resolveBaseline(carbon)
  const bd = carbon.breakdown || {}
  let actual = num(carbon.actual_cost_gco2e)
  if (actual <= 0) actual = num(bd.actual_co2e_g)
  if (actual <= 0) {
    const actEnergy =
      num(carbon.actual_energy_kwh) || num(bd.optimized_energy_kwh)
    const intensity =
      num(carbon.local_grid_gco2_kwh) ||
      num(bd.grid_carbon_intensity_gco2_kwh)
    if (actEnergy > 0 && intensity > 0) actual = actEnergy * intensity
  }

  if (baseline <= 0 && actual <= 0) return null

  // Signed savings — do not clamp to zero
  let saved: number
  if (carbon.carbon_saved_grams != null || bd.carbon_saved_g != null) {
    saved = num(carbon.carbon_saved_grams ?? bd.carbon_saved_g)
  } else {
    saved = baseline - actual
  }
  let efficiency: number
  if (carbon.efficiency_percent != null || bd.reduction_percent != null) {
    efficiency = num(carbon.efficiency_percent ?? bd.reduction_percent)
  } else {
    efficiency = baseline > 0 ? (saved / baseline) * 100 : 0
  }

  const comparison_models: ComparisonModelRow[] = FRONTIER_MODEL_J_PER_TOKEN.map(
    ([model, modelJ]) => {
      const estimated = estimateFrontierG(carbon, baseline, modelJ)
      const savedG = estimated - actual
      const reduction = estimated > 0 ? (savedG / estimated) * 100 : 0
      return {
        model,
        relative_factor: Math.round((modelJ / A_J_PER_TOKEN_MEDIUM) * 10000) / 10000,
        estimated_gco2e: round1(estimated),
        saved_gco2e: round1(savedG),
        reduction_percent: round1(reduction),
      }
    },
  )

  const badges: string[] = []
  const seen = new Set<number>()
  ;[...comparison_models]
    .sort(
      (a, b) =>
        b.reduction_percent - a.reduction_percent ||
        b.saved_gco2e - a.saved_gco2e,
    )
    .forEach((row) => {
      if (badges.length >= 3) return
      if (seen.has(row.relative_factor)) return
      const pct = Math.round(row.reduction_percent)
      if (pct <= 0) return
      seen.add(row.relative_factor)
      badges.push(`${pct}% less CO₂ than ${row.model}`)
    })

  const chart_bars = [
    ...comparison_models.map((row) => ({
      model: row.model,
      estimated_gco2e: row.estimated_gco2e,
      is_ours: false,
    })),
    {
      model: OUR_SYSTEM_NAME,
      estimated_gco2e: round1(actual),
      is_ours: true,
    },
  ].sort((a, b) => b.estimated_gco2e - a.estimated_gco2e)

  const chunk_breakdown =
    carbon.chunk_breakdown || bd.chunk_breakdown || []

  return {
    comparison_models,
    our_system: {
      name: OUR_SYSTEM_NAME,
      tagline: OUR_SYSTEM_TAGLINE,
      carbon: round1(actual),
    },
    summary_cards: {
      actual_emissions_gco2e: round4(actual),
      carbon_saved_gco2e: round4(saved),
      reduction_percent: round1(efficiency),
      heavy_model_baseline_gco2e: round4(baseline),
      estimated_optimized_pipeline_emissions_g: round4(actual),
      estimated_baseline_pipeline_emissions_g: round4(baseline),
      emissions_direction:
        carbon.emissions_direction || bd.emissions_direction || undefined,
    },
    badges,
    chart_bars,
    chunk_breakdown,
    methodology:
      carbon.methodology ||
      "Each frontier bar = entire document workflow on that single model. Our system = per-chunk Light/Medium/Heavy routing. Baseline = naive all-heavy frontier.",
  }
}

/** Always rebuild from carbon_data so the UI cannot show stale 252g bars. */
export function resolveFrontierComparison(
  result:
    | {
        carbon_data?: CarbonLike | null
        comparison_models?: ComparisonModelRow[] | null
        our_system?: { name: string; tagline?: string; carbon: number } | null
        summary_cards?: FrontierComparisonPayload["summary_cards"] | null
        badges?: string[] | null
        chart_bars?: FrontierComparisonPayload["chart_bars"] | null
        methodology?: string | null
      }
    | null
    | undefined,
): FrontierComparisonPayload | null {
  if (!result) return null
  const rebuilt = buildFrontierComparisonFromCarbon(result.carbon_data)
  if (rebuilt) return rebuilt
  if (!result.comparison_models?.length) return null
  return {
    comparison_models: result.comparison_models,
    our_system: {
      name: result.our_system?.name || OUR_SYSTEM_NAME,
      tagline: result.our_system?.tagline || OUR_SYSTEM_TAGLINE,
      carbon: num(result.our_system?.carbon),
    },
    summary_cards: result.summary_cards || {
      actual_emissions_gco2e: 0,
      carbon_saved_gco2e: 0,
      reduction_percent: 0,
      heavy_model_baseline_gco2e: 0,
    },
    badges: result.badges || [],
    chart_bars: result.chart_bars || [],
    methodology: result.methodology || "",
  }
}

File: backend/src/core/frontier_carbon_compare.py
"""
Frontier-model comparison visualization (Boundary A).

Each frontier row estimates operational CO₂e if the *entire document workflow*
(same map + compile token mass + shared stages) were served by that single
model — identical workload to the naive heavy baseline, only J/token changes.

Our system bar uses the smart-routed ``actual_cost_gco2e``.
"""
from __future__ import annotations

from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple

from src.carbon import assumptions as A
from src.carbon.accounting import METHODOLOGY_TEXT as CARBON_METHODOLOGY
from src.carbon.energy_model import energy_to_co2e_g, joules_to_kwh

FRONTIER_MODEL_J_PER_TOKEN: Tuple[Tuple[str, float], ...] = (
    ("GPT-o3", 7.5),
    ("GPT-4", A.J_PER_TOKEN_TYPICAL["heavy"]),
    ("Llama 4 Behemoth", A.J_PER_TOKEN_TYPICAL["heavy"]),
    ("Claude 4 Opus", 6.0),
    ("GPT-4o", 4.0),
    ("Gemini 2.5 Pro", 3.8),
    ("Llama 4 Maverick", 2.8),
    ("Gemma 3", max(A.J_PER_TOKEN_TYPICAL["light"], 1.6)),
)

FRONTIER_RELATIVE_INTENSITY: Tuple[Tuple[str, float], ...] = tuple(
    (name, round(j / A.J_PER_TOKEN_TYPICAL["medium"], 4))
    for name, j in FRONTIER_MODEL_J_PER_TOKEN
)

OUR_SYSTEM_NAME = "Green Agentic Document Processing System"
OUR_SYSTEM_TAGLINE = "Smart Carbon-Aware Routing"

COMPARISON_METHODOLOGY = (
    "Each frontier bar answers: what if the ENTIRE document workflow "
    "(same map + compile token mass and shared parse/chunk/embed/retrieve stages) "
    "were processed by that single model? "
    "CO₂e = (shared_compute + (map+compile)×J_model) × PUE / 3.6e6 × Electricity Maps intensity. "
    "Our system bar is the carbon-aware per-chunk routed pipeline. "
    "Baseline reference = naive all-heavy frontier. "
    + CARBON_METHODOLOGY
)

METHODOLOGY_TEXT = COMPARISON_METHODOLOGY
BADGE_TOP_N = 3


def _round1(value: float) -> float:
    return round(float(value), 1)


def _round4(value: float) -> float:
    return round(float(value), 4)


def _safe_float(value: Any, default: float = 0.0) -> float:
    try:
        if value is None:
            return default
        return float(value)
    except (TypeError, ValueError):
        return default


def _resolve_baseline_co2(data: Mapping[str, Any]) -> float:
    breakdown = data.get("breakdown") if isinstance(data.get("breakdown"), dict) else {}
    baseline = _safe_float(data.get("baseline_cost_gco2e"))
    if baseline <= 0:
        baseline = _safe_float(breakdown.get("baseline_co2e_g"))
    if baseline > 0:
        return baseline
    energy_kwh = _safe_float(data.get("baseline_energy_kwh"))
    if energy_kwh <= 0:
        energy_kwh = _safe_float(breakdown.get("baseline_energy_kwh"))
    intensity = _safe_float(data.get("local_grid_gco2_kwh"))
    if intensity <= 0:
        intensity = _safe_float(breakdown.get("grid_carbon_intensity_gco2_kwh"))
    if energy_kwh > 0 and intensity > 0:
        return energy_kwh * intensity
    return 0.0


def _workflow_inference_tokens(breakdown: Mapping[str, Any], data: Mapping[str, Any]) -> int:
    """Map + compile token mass — same workload as the naive frontier baseline."""
    map_total = int(_safe_float(breakdown.get("map_tokens_total")))
    if map_total <= 0:
        by_tier = breakdown.get("map_tokens_by_tier") or {}
        if isinstance(by_tier, dict):
            map_total = sum(int(_safe_float(v)) for v in by_tier.values())
    compile_tok = int(
        _safe_float(breakdown.get("compile_tokens") or data.get("compile_tokens"))
    )
    if map_total + compile_tok > 0:
        return map_total + compile_tok
    inp = int(_safe_float(breakdown.get("input_tokens") or data.get("input_tokens")))
    if inp > 0:
        return max(int(inp * 1.25), inp) + max(inp // 3, 0)
    return 0


def _estimate_frontier_gco2e(
    *,
    model_j_per_token: float,
    baseline_g: float,
    breakdown: Mapping[str, Any],
    data: Mapping[str, Any],
) -> float:
    intensity = _safe_float(data.get("local_grid_gco2_kwh"))
    if intensity <= 0:
        intensity = _safe_float(breakdown.get("grid_carbon_intensity_gco2_kwh"))
    if intensity <= 0 or baseline_g <= 0:
        return 0.0

    stages = breakdown.get("baseline_stages_gco2e")
    if not isinstance(stages, dict):
        stages = {}

    inference_g = _safe_float(stages.get("inference_gco2e"))
    other_g = 0.0
    for key, val in stages.items():
        if key in ("inference_gco2e", "total_gco2e", "infrastructure_gco2e"):
            continue
        other_g += _safe_float(val)
    infra_g = _safe_float(stages.get("infrastructure_gco2e"))

    if inference_g <= 0:
        inference_g = baseline_g * 0.86
        other_g = baseline_g * 0.05
        infra_g = max(0.0, baseline_g - inference_g - other_g)

    token_mass = _workflow_inference_tokens(breakdown, data)
    if token_mass <= 0:
        heavy_j = A.J_PER_TOKEN_TYPICAL["heavy"]
        ratio = float(model_j_per_token) / heavy_j if heavy_j > 0 else 1.0
        new_inference = inference_g * ratio
        it_new = other_g + new_inference
        it_old = other_g + inference_g
        infra_new = infra_g * (it_new / it_old) if it_old > 0 else infra_g
        return max(0.0, it_new + infra_new)

    inference_j = float(token_mass) * float(model_j_per_token)
    # Match stage attribution: IT grams without PUE; infrastructure carries (PUE−1).
    new_inference_g = energy_to_co2e_g(joules_to_kwh(inference_j), intensity)

    it_old = other_g + inference_g
    it_new = other_g + new_inference_g
    infra_new = infra_g * (it_new / it_old) if it_old > 0 else infra_g
    return max(0.0, it_new + infra_new)


def build_frontier_comparison(
    carbon_data: Optional[Mapping[str, Any]],
    *,
    model_factors: Optional[Sequence[Tuple[str, float]]] = None,
    badge_top_n: int = BADGE_TOP_N,
) -> Dict[str, Any]:
    data = dict(carbon_data or {})
    baseline = _resolve_baseline_co2(data)
    actual = _safe_float(data.get("actual_cost_gco2e"))
    breakdown = data.get("breakdown") if isinstance(data.get("breakdown"), dict) else {}
    if actual <= 0:
        actual = _safe_float(breakdown.get("actual_co2e_g"))
    if "carbon_saved_grams" in data:
        saved = _safe_float(data.get("carbon_saved_grams"))
    else:
        saved = baseline - actual if baseline > 0 else 0.0
    if "efficiency_percent" in data:
        efficiency = _safe_float(data.get("efficiency_percent"))
    elif baseline > 0:
        efficiency = (saved / baseline) * 100.0
    else:
        efficiency = 0.0

    if baseline > 150:
        energy_kwh = _safe_float(data.get("baseline_energy_kwh")) or _safe_float(
            breakdown.get("baseline_energy_kwh")
        )
        intensity = _safe_float(data.get("local_grid_gco2_kwh")) or _safe_float(
            breakdown.get("grid_carbon_intensity_gco2_kwh")
        )
        if energy_kwh > 0 and intensity > 0:
            rebuilt = energy_kwh * intensity
            if 0 < rebuilt < baseline:
                baseline = rebuilt

    if model_factors is not None:
        medium_j = A.J_PER_TOKEN_TYPICAL["medium"]
        models: List[Tuple[str, float]] = [
            (name, float(factor) * medium_j) for name, factor in model_factors
        ]
    else:
        models = list(FRONTIER_MODEL_J_PER_TOKEN)

    comparison_models: List[Dict[str, Any]] = []
    medium_j = A.J_PER_TOKEN_TYPICAL["medium"]

    for model_name, model_j in models:
        estimated = _estimate_frontier_gco2e(
            model_j_per_token=float(model_j),
            baseline_g=baseline,
            breakdown=breakdown,
            data=data,
        )
        saved_g = estimated - actual
        reduction = (saved_g / estimated) * 100.0 if estimated > 0 else 0.0
        comparison_models.append(
            {
                "model": model_name,
                "relative_factor": round(float(model_j) / medium_j, 4) if medium_j else 1.0,
                "j_per_token": float(model_j),
                "estimated_gco2e": _round1(estimated),
                "saved_gco2e": _round1(saved_g),
                "reduction_percent": _round1(reduction),
            }
        )

    ranked = sorted(
        comparison_models,
        key=lambda row: (row["reduction_percent"], row["saved_gco2e"]),
        reverse=True,
    )
    badges: List[str] = []
    seen_factors: set = set()
    for row in ranked:
        if len(badges) >= max(0, int(badge_top_n)):
            break
        factor = row.get("relative_factor")
        if factor in seen_factors:
            continue
        pct = int(round(row["reduction_percent"]))
        if pct <= 0:
            continue
        seen_factors.add(factor)
        badges.append(f"{pct}% less CO₂ than {row['model']}")

    chart_bars: List[Dict[str, Any]] = [
        {
            "model": row["model"],
            "estimated_gco2e": row["estimated_gco2e"],
            "is_ours": False,
        }
        for row in comparison_models
    ]
    chart_bars.append(
        {
            "model": OUR_SYSTEM_NAME,
            "estimated_gco2e": _round1(actual),
            "is_ours": True,
        }
    )
    chart_bars.sort(key=lambda row: row["estimated_gco2e"], reverse=True)

    return {
        "comparison_models": comparison_models,
        "our_system": {
            "name": OUR_SYSTEM_NAME,
            "tagline": OUR_SYSTEM_TAGLINE,
            "carbon": _round1(actual),
        },
        "summary_cards": {
            "actual_emissions_gco2e": _round4(actual),
            "carbon_saved_gco2e": _round4(saved),
            "reduction_percent": _round1(efficiency),
            "heavy_model_baseline_gco2e": _round4(baseline),
            "estimated_optimized_pipeline_emissions_g": _round4(actual),
            "estimated_baseline_pipeline_emissions_g": _round4(baseline),
            "emissions_direction": data.get("emissions_direction")
            or breakdown.get("emissions_direction"),
            "reporting_boundary_label": data.get("reporting_boundary_label")
            or "Operational Emissions (Boundary A)",
        },
        "badges": badges,
        "chart_bars": chart_bars,
        "methodology": COMPARISON_METHODOLOGY,
        "breakdown": breakdown,
        "chunk_breakdown": data.get("chunk_breakdown")
        or breakdown.get("chunk_breakdown")
        or [],
    }
